components/dashboard/src/components/RepositoryFinder.tsx

```diff
@@ -0,0 +1,159 @@
+        if (JSON.stringify(results) !== JSON.stringify(searchResults)) {
+            setSearchResults(results);
+            setSelectedSearchResult(results[0]);
+        }
+    }
+
+    const search = async (query: string) => {
+        setSearchQuery(query);
+        await findResults(query, onResults);
+        if (await refreshSearchData(query, user)) {
+            // Re-run search if the underlying search data has changed
+            await findResults(query, onResults);
+        }
+    }
+
+    // Up/Down keyboard navigation between results
+    const onKeyDown = (event: React.KeyboardEvent) => {
+        if (!selectedSearchResult) {
+            return;
+        }
+        const selectedIndex = searchResults.indexOf(selectedSearchResult);
+        const select = (index: number) => {
+            // Implement a true modulus in order to "wrap around" (e.g. `select(-1)` should select the last result)
+            // Source: https://stackoverflow.com/a/4467559/3461173
+            const n = Math.min(searchResults.length, MAX_DISPLAYED_ITEMS);
+            setSelectedSearchResult(searchResults[((index % n) + n) % n]);
+        }
+        if (event.key === 'ArrowDown') {
+            event.preventDefault();
+            select(selectedIndex + 1);
+            return;
+        }
+        if (event.key === 'ArrowUp') {
+            event.preventDefault();
+            select(selectedIndex - 1);
+            return;
+        }
+    }
+
+    useEffect(() => {
+        const element = document.querySelector(`a[href='/#${selectedSearchResult}']`);
+        if (element) {
+            element.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
+        }
+    }, [selectedSearchResult]);
+
+    const onSubmit = (event: React.FormEvent) => {
+        event.preventDefault();
+        if (selectedSearchResult) {
+            window.location.href = '/#' + selectedSearchResult;
+        }
+    }
+
+    return <form onSubmit={onSubmit}>
+        <div className="flex px-4 rounded-xl border border-gray-300 dark:border-gray-500">
+            <div className="py-4">
+                <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 16 16" width="16" height="16"><path fill="#A8A29E" d="M6 2a4 4 0 100 8 4 4 0 000-8zM0 6a6 6 0 1110.89 3.477l4.817 4.816a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 010 6z" /></svg>
+            </div>
+            <input type="search" className="flex-grow" placeholder="Repository" autoFocus value={searchQuery} onChange={e => search(e.target.value)} onKeyDown={onKeyDown} />
+        </div>
+        <div className="mt-3 flex flex-col space-y-2 h-64 overflow-y-auto" id="search-results">
+            {searchResults.slice(0, MAX_DISPLAYED_ITEMS).map((result, index) =>
+                <a className={`px-4 py-3 rounded-xl` + (result === selectedSearchResult ? ' bg-gray-600 text-gray-50 dark:bg-gray-700' : '')} href={`/#${result}`} key={`search-result-${index}`} onMouseEnter={() => setSelectedSearchResult(result)}>
```

Comment:
**suggestion**: It would be probably good to add some right margin here to avoid squashing the highlighted reasult entry with the visible scrollbar. What do you think?

```suggestion
                <a className={`px-4 py-3 rounded-xl mr-2` + (result === selectedSearchResult ? ' bg-gray-600 text-gray-50 dark:bg-gray-700' : '')} href={`/#${result}`} key={`search-result-${index}`} onMouseEnter={() => setSelectedSearchResult(result)}>
```

| BEFORE | AFTER |
|-|-|
| <img width="1440" alt="Screenshot 2022-01-21 at 6 17 21 PM (2)" src="https://user-images.githubusercontent.com/120486/150561928-8b64e57a-a944-4a74-92e7-d2b08110d27f.png"> | <img width="1440" alt="Screenshot 2022-01-21 at 6 17 35 PM (2)" src="https://user-images.githubusercontent.com/120486/150561973-bd68a2d9-84ec-4e45-beaa-71a0590ea9c0.png"> |

Reply:
Uh oh 😬 I'm not a big fan of either, and I think I'd prefer having the results be the same width as the input. I'll see if I can improve this somehow.

Reply:
Yeah, [no biggie](https://en.wiktionary.org/wiki/no_biggie). Regarding width, same width would be nice but that could require us to move the scrollbar outside the scrolling area. Posting below a slightly related design with scrollbars from https://github.com/gitpod-io/gitpod/issues/4948 to better visualize the problem. In any case, feel free to leave this as is, too. 🏓 

<img width="1440" alt="127672486-fae3178f-e301-433d-9275-0e198e2a35ff" src="https://user-images.githubusercontent.com/120486/150782333-cfb07363-dead-4dc0-97c3-49f0a493e610.png">

Reply:
@gtsiolis Okay, I believe I've found an elegant fix for this 😁 I've simply made the scroll zone wider (negative X-margin) and compensated with some padding (positive X-padding). This doesn't require changing the size of other elements, and has the effect of "simply moving the scrollbar a bit to the right":

<img width="972" alt="Screenshot 2022-02-01 at 15 40 55" src="https://user-images.githubusercontent.com/599268/151989919-0d4be4ea-a8b3-4e2d-846d-24bea875e433.png">
<img width="440" alt="Screenshot 2022-02-01 at 15 43 17" src="https://user-images.githubusercontent.com/599268/151989961-c5eb622c-f987-497b-919a-63612ec9c645.png">

Reply:
Thanks for giving this a go, @jankeromnes!

**suggestion**: I think going with the simple inner right padding for now sounds better.

**issue**: Moving the scrollbar outside could make the user feel like they can scroll the through the whole modal inner elements, including title, text input, etc, not just the repository list. This also becomes more visible or an issue when always showing scrollbars, see screenshots below. This is also what I see! 👀 

| Light | Dark |
|-|-|
| <img width="528" alt="scrollbar-light" src="https://user-images.githubusercontent.com/120486/152153013-25df8108-d644-4ad6-aa40-08ada44886c8.png"> | <img width="528" alt="scrollbar-dark" src="https://user-images.githubusercontent.com/120486/152153016-38d1d997-e2b3-4a8d-900f-646f533745e4.png"> |

**thought**: Keeping the scrollbar near the scrollbar area could help us :a: use better layouts like the one with the gray background mentioned in https://github.com/gitpod-io/gitpod/pull/7715#discussion_r793756071, :b: potentially introduce more fixed elements like a bottom toolbar for surfacing results found, etc.

Regarding the scrollbar blindness issue we discussed a few days ago, here's a relevant thread[[1](https://news.ycombinator.com/item?id=24293421)] but unfortunately the original article has been removed. I could only find this partial screenshot[[2](https://www.are.na/block/8469033)]. 🌟